gen/src/main.rs

```diff
@@ -0,0 +1,130 @@
+extern crate getopts;
+
+extern crate futures;
+extern crate hyper;
+extern crate tokio_core;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate serde_yaml;
+
+#[macro_use]
+extern crate lazy_static;
+
+#[macro_use]
+extern crate matches;
+extern crate regex;
+
+mod download;
+mod generate;
+
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+use download::download;
+use generate::generate_all;
+
+use getopts::Options;
+
+fn main() {
+    let mut opts = Options::new();
+
+    opts.optflag("h", "help", "Print this help menu.");
+    opts.optopt(
+        "u",
+        "unicode-version",
+        "The version of the unicode standard to download.  \
+         If omitted, the local cache will be used.",
+        "VERSION",
+    );
+    opts.optmulti(
+        "",
+        "crate",
+        "A crate to generate tables for.  \
+         This propogates through the tree: if you specify only `unic`, all crates will be built.",
+        "CRATE",
+    );
+
+    let matches = opts.parse(env::args().skip(1))
+        .unwrap_or_else(|e| panic!(e.to_string()));
+
+    if matches.opt_present("h") {
+        print_usage(&opts);
+        return;
+    }
+
+    matches.opt_str("u").map(|version| {
+        println!("Downloading Unicode resources version {}...", version);
+        download(version.as_str()).expect("Failed to download Unicode resources");
+        println!("Finished.");
+    });
+
+    let crates = expand_sub_crates(matches.opt_strs("crate").into_iter());
+
+    if crates.is_empty() {
+        if !matches.opt_present("u") {
+            print_usage(&opts);
+        }
+        return;
+    }
+
+    println!("Generating sources for {} crates...", crates.len());
+    generate_all(crates).expect("Failed to generate sources");
+    println!("Finished.");
+}
+
+/// Path to crate disassembly mapping (YAML).
+///
+/// Format:
+///
+/// ```yaml
+/// ---
+/// crate:
+///   - "crate-subcrate1"
+///   - "crate-subcrate2"
+/// crate-subcrate1:
+///   - "crate-subcrate1-subcrate"
+/// ```
+const CRATES: &'static str = "gen/config/crates.yaml";
+
+/// Take a crate name in the UNIC hierarchy and expand it into all subcrates.
+///
+/// Reads `config/crates.yaml` for the crate disassembly process.
+///
+/// Technically, a crate only needs to be in that mapping if it needs to have sources built.
+/// For consistency, all crates that need generated sources or not should be passed into the
+/// generation module, which will ignore crates it doesn't need to do anything for.
+fn expand_sub_crates<I>(list: I) -> Vec<String>
```

Comment:
I'm going to need a bit more of a nudge to understand exactly what you're suggesting here.

Focusing on usage, my current idea of how the config will be laid out (if you'll excuse the vague syntax, I've not familiarized myself with the ins and outs of TOML yet):

```text
{
  idna: {
    version: "10.0.0"
    resources: [
      { url: "http://www.unicode.org/Public/idna/{version}/IdnaMappingTable.txt"
        dest: "data/idna/IdnaMappingTable.txt" }
      { url: "http://www.unicode.org/Public/idna/{version}/ReadMe.txt"
        dest: "data/idna/ReadMe.txt" }
      { url: "http://www.unicode.org/Public/idna/{version}/IdnaTest.txt"
        dest: "data/idna/test/IdnaTest.txt" }
    ]
  }
  ucd: {
    version: "10.0.0"
    resources: [
      { url: "http://www.unicode.org/Public/{}/ucd/DerivedAge.txt"
        dest: "data/ucd/DerivedAge.txt" }
    ]
  }
}
```

(and so on)

So then usage would then be

```bash
$ cargo run --crate=unic-gen -- 
# prints help

$ cargo run --crate=unic-gen -- --help
# prints help

$ cargo run --crate=unic-gen -- --download-resources
# downloads all resources

$ cargo run --crate=unic-gen -- --target idna --download-resources
# downlaods resources for idna

$ cargo run --crate=unic-gen -- --generate-tables
# generates all tables

$ cargo run --crate=unic-gen -- --target ucd --generate-tables
# generates tables for ucd
```

So the `expand_sub_crates` fn would be unnecessary by the fact that we only care about the first-level subcrates, and there is no need to expand a request for `unic-ucd` into `unic-ucd-core`, `unic-ucd-age`, etc.